File: Coding Sample/MS.NET/Fundamentals/Language/OOP/banking1.cs
using System;

namespace Banking
{
	public enum FaultType {InsufficientFunds, IllegalTransfer}

	public class AccountingException : ApplicationException
	{
		public FaultType Cause {get;}

		public AccountingException(FaultType fault) : base(fault.ToString())
		{
			this.Cause = fault;
		}
	}

	public abstract class Account
	{
		public string Id {get; internal set;}

		public double Balance {get; protected set;}

		public abstract void Deposit(double amount);

		public abstract void Withdraw(double amount);

		public void Transfer(double amount, Account that)
		{
			if(ReferenceEquals(this, that))
				throw new AccountingException(FaultType.IllegalTransfer);

			this.Withdraw(amount);
			that.Deposit(amount);
		}
	}

	sealed class CurrentAccount : Account
	{
		public override void Deposit(double amount)
		{
			base.Balance += amount;
		}

		public override void Withdraw(double amount)
		{
			base.Balance -= amount;
		}
	}

	sealed class SavingsAccount : Account
	{
		const double MinBalance = 5000;

		internal SavingsAccount()
		{
			base.Balance = MinBalance;
		}

		public override void Deposit(double amount)
		{
			base.Balance += amount;
		}

		public override void Withdraw(double amount)
		{
			if(base.Balance - amount < MinBalance)
				throw new AccountingException(FaultType.InsufficientFunds);

			base.Balance -= amount;
		}
	}

	public static class Banker
	{
		private static int nid = Environment.TickCount % 1000000;

		public static Account OpenCurrentAccount()
		{
			CurrentAccount acc = new CurrentAccount();

			acc.Id = "C/A" + nid++;

			return acc;
		}

		public static Account OpenSavingsAccount()
		{
			SavingsAccount acc = new SavingsAccount();

			acc.Id = "S/A" + nid++;

			return acc;
		}
	}
}

File: Coding Sample/Foundations/Advanced/C++/Inheritance/banking1.cpp
#include "banking1.h"

namespace Banking
{
	long Account::Id() const
	{
		return id;
	}

	double Account::Balance() const
	{
		return balance;
	}

	void Account::Transfer(double amount, Account* that) throw(InsufficientFunds)
	{
		if(this != that)
		{
			this->Withdraw(amount);
			that->Deposit(amount);
		}
	}

	Account::Account()
	{
		static long nid = std::time(0) % 10000;

		id = nid++;
		balance = 0;
	}

	void CurrentAccount::Deposit(double amount)
	{
		if(balance < 0)
			amount *= 0.9;
		balance += amount;
	}

	void CurrentAccount::Withdraw(double amount) throw(InsufficientFunds)
	{
		balance -= amount;
	}

	SavingsAccount::SavingsAccount()
	{
		balance = 5000;
	}

	void SavingsAccount::Deposit(double amount)
	{
		balance += amount;
	}

	void SavingsAccount::Withdraw(double amount) throw(InsufficientFunds)
	{
		if(balance - amount < 5000)
			throw InsufficientFunds();

		balance -= amount;
	}

	double SavingsAccount::AddInterest(short period)
	{
		float rate = balance < 10000 ? 4 : 5;
		double interest = balance * period * rate / 100;

		balance += interest;

		return interest;
	}
}

